### PR TITLE
Entitlement fixes

### PR DESCRIPTION
**Pull request by pierre**
See https://github.com/killbill/killbill-integration-tests/pull/27.

### CODE REVIEW

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java [hunk 1]
```diff
@@ -506,8 +372,184 @@ public UUID doCall(final EntitlementApi entitlementApi, final EntitlementContext
...
+
+        final Iterator<BaseEntitlementWithAddOnsSpecifier> iterator = baseEntitlementWithAddOnsSpecifiers.iterator();
+        if (!iterator.hasNext()) {
+            throw new SubscriptionBaseApiException(ErrorCode.SUB_CREATE_INVALID_ENTITLEMENT_SPECIFIER);
+        }
+
+        return iterator.next();
+    }
+
+    private void preCheckAddEntitlement(final BaseEntitlementWithAddOnsSpecifier baseEntitlementWithAddOnsSpecifier,
```
Comment (sbrossie) [inline entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java]: Nit: This is slightly more than a preCheck as it also populate`populateEventsStreamForBaseSubscriptionPerBundleCache ` -- and  cache.

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java [hunk 2]
```diff
@@ -506,8 +372,184 @@ public UUID doCall(final EntitlementApi entitlementApi, final EntitlementContext
...
+                final EventsStream eventsStreamForBaseSubscription = eventsStreamBuilder.buildForBaseSubscription(bundleId, callContext);
+                eventsStreamForBaseSubscriptionPerBundle.put(bundleId, Optional.<EventsStream>of(eventsStreamForBaseSubscription));
+            } else {
+                eventsStreamForBaseSubscriptionPerBundle.put(bundleId, Optional.<EventsStream>absent());
+            }
+        }
+    }
+
+    private void preCheckAddEntitlement(final UUID bundleId, final DateTime entitlementRequestedDate, final BaseEntitlementWithAddOnsSpecifier baseEntitlementWithAddOnsSpecifier, final EventsStream eventsStreamForBaseSubscription) throws EntitlementApiException {
+        if (eventsStreamForBaseSubscription.isEntitlementCancelled() ||
```
Comment (sbrossie) [inline entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java]: In theory we should check the cancellation date with respect to the `entitlementRequestedDate` and allow if such cancellation date is more in the future.

I think this goes into the discussion we had the other day with my fix for usage record. Ideally we would have a method which given an effective date a a subscription checks for validity. At a high level something like:
* If `requestedDate` < `startDate` -> Denied
* If `requestedDate` >= `cancelDate` -> Denied
* If `requestedDate` is during time period where subscription is not Active -> denied

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java [hunk 3]
```diff
@@ -506,8 +372,184 @@ public UUID doCall(final EntitlementApi entitlementApi, final EntitlementContext
...
+                                        final Catalog catalog,
+                                        final TenantContext callContext,
+                                        final InternalCallContext contextWithValidAccountRecordId) throws EntitlementApiException {
+        // In the addEntitlement codepath, bundleId is always set. But, technically, an existing bundle could be specified by externalKey in
+        // the createBaseEntitlementsWithAddOns codepath. In that case, we should also check if that bundle is blocked.
+        UUID bundleId = baseEntitlementWithAddOnsSpecifier.getBundleId();
+        if (bundleId == null && baseEntitlementWithAddOnsSpecifier.getExternalKey() != null) {
+            populateBundleKeyToIdMappingCache(baseEntitlementWithAddOnsSpecifier, bundleKeyToIdMapping, catalog, contextWithValidAccountRecordId);
+
+            final Optional<UUID> bundleIdForKey = bundleKeyToIdMapping.get(baseEntitlementWithAddOnsSpecifier.getExternalKey());
```
Comment (sbrossie) [inline entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java]: Nit: Are we embracing `Optional` inside KB core?

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java (hunk 1 above)
Comment (pierre) [reply entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java]: Fixed in 052a6a9dcdd525870f04b37fa6755a3154c73641.

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java (hunk 2 above)
Comment (pierre) [reply entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java]: I believe this is already in place?

Here's the full condition:

```
if (eventsStreamForBaseSubscription.isEntitlementCancelled() ||
     eventsStreamForBaseSubscription.isEntitlementPending() &&
     (baseEntitlementWithAddOnsSpecifier.getEntitlementEffectiveDate() == null ||
              baseEntitlementWithAddOnsSpecifier.getEntitlementEffectiveDate().compareTo(eventsStreamForBaseSubscription.getEntitlementEffectiveStartDate()) < 0))) {
            throw new EntitlementApiException(ErrorCode.SUB_GET_NO_SUCH_BASE_SUBSCRIPTION, bundleId);
}
```

Note that `isEntitlementCancelled` means we are past the effective cancellation date.

**Pull request merged**